### PR TITLE
Reload tensorflow model with an RPC call

### PR DESCRIPTION
**Pull request by rajhans**
This PR does the following
1) Creates an AdminService to reload the model.
2) The services uses a pointer to a unique pointer from PredictionService. 
3) Added test to tensorflow_model_server_test to check that ReloadWorksCorrectly.

### CODE REVIEW

**Comment by pratyus** [comment]
Can you please add more details in this PR?

File: tensorflow_serving/apis/prediction_service.proto [hunk 1]
```diff
@@ -8,6 +8,7 @@ import "tensorflow_serving/apis/get_model_metadata.proto";
 import "tensorflow_serving/apis/inference.proto";
 import "tensorflow_serving/apis/predict.proto";
 import "tensorflow_serving/apis/regression.proto";
+import "tensorflow_serving/config/model_server_config.proto";
 
 // open source marker; do not remove
 // PredictionService provides access to machine-learned models loaded by
```
Comment (n-rook) [inline tensorflow_serving/apis/prediction_service.proto]: All the PredictionService API methods don't modify state, but your new one does. I would not be surprised if the TF folks ask you to put it in a new service, like an AdminService or ConfigService something like that. I wouldn't bother doing that ahead of time, though.

File: tensorflow_serving/apis/prediction_service.proto [hunk 2]
```diff
@@ -22,6 +23,9 @@ service PredictionService {
   // Predict -- provides access to loaded TensorFlow model.
   rpc Predict(PredictRequest) returns (PredictResponse);
 
+  // Reload -- reloads modesl from config.
+  rpc Reload(ModelServerConfig) returns (PredictResponse);
```
Comment (n-rook) [inline tensorflow_serving/apis/prediction_service.proto]: All of the other API calls use unique request/response protos, so you'll want to do so too.

File: tensorflow_serving/example/mnist_client.py [hunk 3]
```diff
@@ -154,6 +156,32 @@ def do_inference(hostport, work_dir, concurrency, num_tests):
         _create_rpc_callback(label[0], result_counter))
   return result_counter.get_error_rate()
```
Comment (n-rook) [inline tensorflow_serving/example/mnist_client.py]: Ignoring this file as requested 😇

**Comment by rajhans** [comment]
This PR is ready for review now.

File: tensorflow_serving/model_servers/main.cc [hunk 4]
```diff
@@ -182,6 +185,24 @@ grpc::Status ToGRPCStatus(const tensorflow::Status& status) {
...
+    VLOG(1) << "Reloading with config " << request->DebugString();
+    const grpc::Status status = ToGRPCStatus((*core_)->ReloadConfig(*request));
+    if (!status.ok()) {
+      VLOG(1) << "Server reload failed: " << status.error_message();
+    }
+    return status;
+  }
+
+ private:
+  std::unique_ptr<ServerCore>* core_;
```
Comment (n-rook) [inline tensorflow_serving/model_servers/main.cc]: I'm no C++ expert, but this strikes me as dangerous. A pointer to a pointer is confusing. If the actual owner of the ServerCore gets destructed, this is going to point to deallocated memory. Could you just use a shared_ptr here instead?

Feel free to ignore, I don't know C++ ;)

File: tensorflow_serving/model_servers/main.cc [hunk 5]
```diff
@@ -274,15 +299,19 @@ class PredictionServiceImpl final : public PredictionService::Service {
   bool use_saved_model_;
 };
 
+
+
```
Comment (n-rook) [inline tensorflow_serving/model_servers/main.cc]: Why all this whitespace?

Comment (rajhans) [reply tensorflow_serving/model_servers/main.cc]: fixed

File: tensorflow_serving/model_servers/main.cc (hunk 4 above)
Comment (rajhans) [reply tensorflow_serving/model_servers/main.cc]: valid point. converted to shared_ptrs everywhere.

**Comment by rajhans** [comment]
And blaze test also passes!

**Comment by 962422135** [comment]
@rajhans With your newly supplied code, I ran the bazel compile error。“config/BUILD:36:1: no such target '//tensorflow_serving/sources/storage_path:file_system_storage_path_source_proto_py_pb2_genproto': target 'file_system_storage_path_source_proto_py_pb2_genproto' not declared in package 'tensorflow_serving/sources/storage_path' defined by /home/tf/workspace/serving/tensorflow_serving/sources/storage_path/BUILD and referenced by '//tensorflow_serving/config:model_server_config_proto_py_pb2_genproto'” (branch is reload-with-grpc)

**Pull request merged**